tests/source.rs

```diff
@@ -0,0 +1,234 @@
+        Regex(regex::Error);
+    }
+
+    links {
+        Rustdoc(rustdoc::error::Error, rustdoc::error::ErrorKind);
+    }
+
+    errors {
+        JsonPointer(p: String) {
+            description("JSON pointer matched no data")
+            display("JSON pointer '{}' matched no data", p)
+        }
+        ValueMismatch(value: String, re: String) {
+            description("the value did not match the regular expression")
+            display("the value '{}' did not match the regular expression '{}'", value, re)
+        }
+        ValueType(value: Value) {
+            description("the JSON pointer did not point at a string")
+            display("the JSON pointer did not point at a string: {:?}", value)
+        }
+    }
+}
+
+/// Create analysis data from a given source file. Returns an analysis host with the data loaded.
+fn generate_analysis(source_file: &Path, tempdir: &Path) -> Result<AnalysisHost> {
+    let source_filename = source_file.to_str().unwrap();
+
+    let rustc_status = Command::new("rustc")
+        .args(&["-Z", "save-analysis"])
+        .arg(source_filename)
+        .current_dir(tempdir.to_str().unwrap())
+        .status()?;
+    if !rustc_status.success() {
+        bail!("compilation of {} failed", source_filename);
+    }
+
+    // rls-analysis expects the analysis files to be in a specific directory -- one usually created
+    // by cargo.
+    let save_analysis_dir = tempdir.join("save-analysis-temp");
+    let expected_analysis_dir = tempdir
+        .join("target")
+        .join("rls")
+        .join(&Target::Debug.to_string())
+        .join("save-analysis");
+    fs::create_dir_all(&expected_analysis_dir)?;
+    fs::rename(&save_analysis_dir, &expected_analysis_dir)?;
+
+    let host = AnalysisHost::new(Target::Debug);
+    host.reload(&tempdir, &tempdir, true)?;
+
+    Ok(host)
+}
+
+/// Runs all tests in a given source file.
+fn check(source_file: &Path, host: &AnalysisHost) -> Result<()> {
+    let package_name = source_file.file_stem().unwrap();
+    let json = DocData::new(host, package_name.to_str().unwrap())?
+        .to_json()?;
+    let json = serde_json::from_str(&json)?;
+
+    let source = BufReader::new(File::open(source_file)?);
+    let mut found_test = false;
+    for line in source.lines() {
+        if let Some(test_case) = parse_test(&line?) {
+            let (pointer, regex) = test_case?;
+            run_test(&json, &pointer, &regex)?;
+            if !found_test {
+                found_test = true;
+            }
+        }
+    }
+
+    if !found_test {
+        bail!("found no tests in {}", source_file.display());
+    }
+
+    Ok(())
+}
+
+/// Optionally parses a test case from a single line. If the line contains a test case, returns a
+/// Result containing a tuple of the JSON pointer and the regular expression. If there is no test
+/// case contained in the line, returns `None`.
+fn parse_test(line: &str) -> Option<Result<(String, Regex)>> {
+    lazy_static! {
```

Comment:
Really? My understanding based off [lazy_statics docs](https://docs.rs/lazy_static/0.2.8/lazy_static/) was that it's put at the top to be reused elsewhere so it can be accessed globally as a static.